File: src/main/java/org/lbogdanov/btlock/objc/ManagerDelegate.java
package org.lbogdanov.btlock.objc;

import static org.lbogdanov.btlock.App.BUS;
import static org.lbogdanov.btlock.objc.Frameworks.CORE_BLUETOOTH;

import org.lbogdanov.btlock.events.PeripheralDiscovered;

import com.apple.jobjc.JObjCRuntime;
import com.apple.jobjc.corebluetooth.CBCentralManager;
import com.apple.jobjc.corebluetooth.CBPeripheral;
import com.apple.jobjc.foundation.NSDictionary;
import com.apple.jobjc.foundation.NSNumber;
import com.apple.jobjc.foundation.NSObject;

public class ManagerDelegate extends NSObject {
    public ManagerDelegate(long objPtr, JObjCRuntime runtime) {
        super(objPtr, runtime);
    }

    public void centralManagerDidUpdateState(CBCentralManager manager) {
        if (manager.state() == CORE_BLUETOOTH.CBCentralManagerStatePoweredOn()) {
            manager.scanForPeripheralsWithServices_options(null, null);
        }
    }

    public void centralManager_didDiscoverPeripheral_advertisementData_RSSI(CBCentralManager manager,
                                                                            CBPeripheral peripheral,
                                                                            NSDictionary data,
                                                                            NSNumber rssi) {
        BUS.post(new PeripheralDiscovered(peripheral));
    }
}

File: src/main/kotlin/btlock/objc/ManagerDelegate.kt
package btlock.objc

import btlock.events.*
import btlock.objc.Frameworks.CORE_BLUETOOTH
import btlock.objc.Frameworks.objcNumbers
import btlock.objc.Frameworks.objcStrings
import com.apple.jobjc.*
import com.apple.jobjc.corebluetooth.*
import com.apple.jobjc.foundation.*
import mu.*
import tornadofx.*

private val logger = KotlinLogging.logger {}

class ManagerDelegateClass : NSObjectClass(Frameworks.JOBJC)

class ManagerDelegate(ptr: Long, rt: JObjCRuntime) : NSObject(ptr, rt) {

    fun centralManagerDidUpdateState(manager: CBCentralManager) {
        if (manager.state() == CORE_BLUETOOTH.CBCentralManagerStatePoweredOn().toLong()) {
            manager.scanForPeripheralsWithServices_options(null, null)
        }
    }

    fun centralManager_didDiscoverPeripheral_advertisementData_RSSI(manager: CBCentralManager,
                                                                    peripheral: CBPeripheral,
                                                                    data: NSDictionary,
                                                                    rssi: NSNumber) {
        val id = objcStrings.javaString(peripheral.identifier().UUIDString())
        logger.info {
            "Peripheral discovered: id=$id, rssi=${objcNumbers.javaNumber(rssi)}"
        }
        FX.eventbus.fire(
            PeripheralDiscovered(id = id, name = objcStrings.javaString(peripheral.name()))
        )
    }

    fun centralManager_didDisconnectPeripheral_error(manager: CBCentralManager,
                                                     peripheral: CBPeripheral,
                                                     error: NSError) {
        logger.info {
            "Peripheral disconnected: id=${objcStrings.javaString(peripheral.identifier().UUIDString())}"
        }
    }

}
